feat(simulation): Add objects for EnergyPlus simulation settings

chriswmackey opened the pull request:
This commit adds all of the objects needed to describe basic parameters of the EnergyPlus simulation including:
* RunPeriod (including start day of week, holidays and daylight savings time)
* Timestep
* ShadowCalculation
* SimulationControl
* SizingParameter

@mostaphaRoudsari , all of the code here is ready for review but I still have to write tests for all of the objects (I will do this soon).

In addition to reviewing the code, I could use your thoughts on one structural decision that I know relates to how things will run on the cloud service.  Right now, the `SimulationParameter` schema (with all of the settings needed for the EnergyPlus simulation) is completely separate for the `Model` schema.  However, we know that at the end of the day, all of this info ends up in the same OSM and IDF. So we could just add `simulation_parameter` as a property for `Model.energy.properties` and then we would have a `Model` schema with a lot of the same info as an IDF. Here are the pros and cons I could think of for either structure:

**Reasons for putting SimulationParameter as a property of the Model**
- JSON files of `Model` objects would have a 1-to-1 relationship with the IDF or OSM
- Only one .json file needs to be parsed by the energy-model-measure to produce a simulate-able OSM (if the `SimulationParameters` are separate from the `Model`, we may need to pass two separate JSON files).

**Reasons for keeping SimulationParameter separate from Model**
- `SimulationParameters` really have more in common with the other assets (.epw, .ddy) than they have with the `Model`. They really aren't really a part of the design being simulated, which is how we have been thinking of the `Model` up to this point.
- In long iterative runs of design options (ie. different `Models`), the `SimulationParameters` aren't likely to change. So it would be useful to only upload them once and not have to upload them for each and every `Model`.
- If we end up making recipes for different types of Energy simulations (ie. a recipe for annual EUI calculation, a recipe for peak loads, a recipe for load balance graphics, a recipe for thermal comfort mapping), it will be useful to have these simulation parameters coordinated with these recipes and not have to edit all of the `Models` for each recipe that we want to use.

mostaphaRoudsari commented:
Hey @chriswmackey, I'm going to review the PR shortly but to answer your question about the relationship between SimulationParameters and Model I think they should stay separate. Model defining a reusable scene which includes the geometry information and meta-data for that geometry.

To me it is OK that models doesn't have a 1-1 relationship with idf files. Keeping the Model separate from parameters for a single run makes the model re-usable and that to me is a key.

chriswmackey commented:
Hey Mostapha. Glad to hear it and I agree that the benefits of keeping them separate outweigh those of having them merged. I will proceed with keeping this PR the way that it is and I will let Tanushree know that she should expect to have two json files to parse in the `energy-model-measure`. Also, there's no rush to review this PR. I'm still working through all of your feedback on the dragonfly-core PR and that should keep me busy for a while.

mostaphaRoudsari commented on honeybee_energy/simulation/sizing.py:
```diff
@@ -0,0 +1,135 @@
+
+    @heating_factor.setter
+    def heating_factor(self, value):
+        self._heating_factor = float_positive(value, 'sizing parameter heating factor')
```
This will fail for value being 0.

chriswmackey replied:
Hm. This type of situation may eventually require its own `typing` method since `float_in_range` won't really work here when 0.000000001 is still a valid input. For now, I'll just add an extra check to make sure it's not 0 but let's consider implementing a new `typing` method if more cases like this crop up in the future.

chriswmackey commented on honeybee_energy/simulation/control.py:
```diff
@@ -0,0 +1,206 @@
+                SimulationControl definition.
+        """
+        # check the inputs
+        ep_strs = parse_idf_string(idf_string, 'SimulationControl,')
```
No it's not needed. I think it's just a good habit since many E+ object names are concatenations of other object names with more words. Putting the comma there always ensures that I have the correct type of IDF string.

chriswmackey commented on honeybee_energy/simulation/control.py:
```diff
@@ -0,0 +1,206 @@
+        Args:
+            data: A SimulationControl dictionary in following the format below.
+
+        .. code-block:: json
```
Ah, you're right. I guess Python and not JSON is what we should be using for all dictionary docstrings, then?

chriswmackey replied:
After seeing Santiago's implementation of `.. code-block:: python` in his latest PRs, I think it's a no-brainer to make all of these `.. code-block:: python`. I'll change this one here and we'll plan to change all of them throughout the library in a later PR.

chriswmackey commented on honeybee_energy/simulation/daylightsavings.py:
```diff
@@ -0,0 +1,191 @@
+        self._check_start_before_end()
+
+    @property
+    def is_leap_year(self):
```
Ok, that's a fair suggestion since you are right that I don't really anticipate people using the `DaylightSavings` object outside of the `Runperiod` object. Let me see if I can change it.

chriswmackey replied:
Ok, I removed this property here and the `RunPeriod` object now handles all coordination of leap year dates.

chriswmackey commented on honeybee_energy/simulation/runperiod.py:
```diff
@@ -0,0 +1,363 @@
+        * daylight_savings
+        * is_leap_year
+    """
+    __slots__ = ('_solar_distribution', '_calculation_method', '_calculation_frequency',
```
Ah, good catch. I guess stupid bugs like this are also the reason why we write tests.

chriswmackey commented on honeybee_energy/simulation/runperiod.py:
```diff
@@ -0,0 +1,363 @@
+        return self.start_date.leap_year
+
+    @is_leap_year.setter
+    def is_leap_year(self, value):
```
You are right that EneryPlus accepts year inputs now in the RunPeriod definition but there are a number of really good reasons why we should not support it:
1. OpenStudio does not support it and, as far as I understand, there are no plans to support it in the future.
2. It would create a coordination nightmare between ladybug and honeybee-energy. To just scratch the surface of the headache that this will create, someone running a multi year energy simulation in E+ will not be able to import their results to ladybug data collections, making it impossible to work with the results in ladybug.
3. If you want to go through the whole effort of exposing year properties across ladybug datetimes and analysis periods to make this work, then you are welcome but this really is an example of letting an edge case drive the core development.  Realistically, the only reason why you'd want to make use of the year input on the RunPeriod is to run a multi-year simulation and, if someone badly needs this, then I don't think it's that big of deal to ask users to just break up their climate data into one year per EPW and run separate simulations.

mostaphaRoudsari replied:
The reason 1 is good enough. I think we wait and see when OpenStudio is going to support this.

chriswmackey commented on honeybee_energy/simulation/runperiod.py:
```diff
@@ -0,0 +1,363 @@
+    @classmethod
+    def from_analysis_period(cls, analysis_period=None, start_day_of_week='Sunday',
+                             holidays=None, daylight_saving_time=None):
+        """Initialize a RunPeriod object from a ladybug AnalysisPeriod.
```
Ok. I'll be more explicit about it in the docstring and add warnings.

chriswmackey commented on honeybee_energy/simulation/shadowcalculation.py:
```diff
@@ -0,0 +1,229 @@
+
+    @calculation_frequency.setter
+    def calculation_frequency(self, value):
+        self._calculation_frequency = int_positive(
```
Ok. I'll change it to use `int_in_range`, with a min of 1, then.

chriswmackey commented:
I finished addressing all of your comments @mostaphaRoudsari and thank you for them (they were all very helpful). I am going to write some tests for all the new code and then we should be able to merge this one in.
